File: android-native/MediaPipePoseModule.kt
/**
 * 2026-05-23 — MediaPipe Pose Landmarker — Android native module.
 *
 * BlazePose 33-keypoint pose detection on-device. Becomes the
 * primary path in services/poseEstimator.ts when present; the cloud
 * /api/pose-analysis route stays the defensive fallback.
 *
 * Public API exposed to React Native:
 *   - detectPoseFromFrame(b64, options, promise)
 *       One-shot pose detection on a single JPEG/PNG frame. Returns
 *       33 normalized keypoints + confidence + world-space landmarks
 *       (when available). This is the path SmartPlay actually uses
 *       today — Phase K extracts 5 keyframes from a swing clip via
 *       expo-video-thumbnails, then calls this once per frame.
 *   - startContinuousDetection(quality, promise)
 *       Subscribes to the running camera (CameraX) and emits
 *       MediaPipePoseFrame events at the model's processing rate.
 *       Reserved for future live-preview use; not wired into any UI
 *       surface today.
 *   - stopContinuousDetection(promise)
 *   - getStatus(promise) → { available, modelLoaded, lastInferenceMs }
 *
 * Quality presets (model variants):
 *   - "lite"  — pose_landmarker_lite.task   (~3 MB, fastest)
 *   - "full"  — pose_landmarker_full.task   (~9 MB, default, what we ship)
 *   - "heavy" — pose_landmarker_heavy.task  (~30 MB, best accuracy)
 *
 * Backward compatibility: never throws on JS-callable methods. Every
 * failure path rejects the promise with a clean DAT_*-style error
 * code; the JS service collapses to "fall back to cloud" without
 * raising to the caller.
 */

package com.smartplaycaddie.mediapipe

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

// MediaPipe imports — resolved by withMediaPipePose.js config plugin
// pulling com.google.mediapipe:tasks-vision into the gradle deps.
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult

class MediaPipePoseModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    @Volatile private var landmarker: PoseLandmarker? = null
    @Volatile private var loadedQuality: String = "full"
    @Volatile private var lastInferenceMs: Long = 0

    override fun getName(): String = "MediaPipePose"

    private fun assetForQuality(quality: String): String = when (quality) {
        "lite"  -> "mediapipe/pose_landmarker_lite.task"
        "heavy" -> "mediapipe/pose_landmarker_heavy.task"
        else    -> "mediapipe/pose_landmarker_full.task"
    }

    /** Lazy init the landmarker. Synchronized so two concurrent
     *  detectPose calls from JS don't double-build. Throws on init
     *  failure — callers wrap in try/promise.reject. */
    @Synchronized
    private fun ensureLandmarker(quality: String) {
        if (landmarker != null && loadedQuality == quality) return
        // Reload only if quality changed.
        landmarker?.close()
        landmarker = null
        val baseOptions = BaseOptions.builder()
            .setModelAssetPath(assetForQuality(quality))
            // Try GPU delegate first — falls back to CPU on devices
            // that don't support GPU acceleration for tasks-vision.
            .setDelegate(Delegate.GPU)
            .build()
        val options = PoseLandmarker.PoseLandmarkerOptions.builder()
            .setBaseOptions(baseOptions)
            .setRunningMode(RunningMode.IMAGE)
            // Single pose — golf is one-player-per-frame for now. If
            // we ever wire a "coach + student" mode, bump this.
            .setNumPoses(1)
            .setMinPoseDetectionConfidence(0.4f)
            .setMinPosePresenceConfidence(0.4f)
            .setMinTrackingConfidence(0.4f)
            .setOutputSegmentationMasks(false)
            .build()
        landmarker = try {
            PoseLandmarker.createFromOptions(reactApplicationContext, options)
        } catch (e: Throwable) {
            // GPU path failed — retry with CPU. Some emulators + older
            // ARM devices don't have the GPU delegate registered.
            val cpuOptions = options.toBuilder()
                .setBaseOptions(
                    BaseOptions.builder()
                        .setModelAssetPath(assetForQuality(quality))
                        .setDelegate(Delegate.CPU)
                        .build(),
                )
                .build()
            PoseLandmarker.createFromOptions(reactApplicationContext, cpuOptions)
        }
        loadedQuality = quality
    }

    @ReactMethod
    fun detectPoseFromFrame(b64: String, options: ReadableMap?, promise: Promise) {
        try {
            val quality = options?.takeIf { it.hasKey("quality") }?.getString("quality") ?: "full"
            ensureLandmarker(quality)
            // Decode base64 → bitmap. Strip data: prefix if present.
            val clean = if (b64.startsWith("data:")) b64.substringAfter(",") else b64
            val bytes = Base64.decode(clean, Base64.DEFAULT)
            val bitmap: Bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                ?: run {
                    promise.reject("MP_DECODE_FAILED", "Could not decode base64 frame to bitmap")
                    return
                }
            val mpImage: MPImage = BitmapImageBuilder(bitmap).build()
            val t0 = System.currentTimeMillis()
            val result: PoseLandmarkerResult = landmarker!!.detect(mpImage)
            lastInferenceMs = System.currentTimeMillis() - t0
            promise.resolve(serializeResult(result, lastInferenceMs))
        } catch (e: Throwable) {
            promise.reject("MP_INFERENCE_FAILED", e.message ?: e.toString())
        }
    }

    @ReactMethod
    fun getStatus(promise: Promise) {
        val map = Arguments.createMap()
        map.putBoolean("available", true)
        map.putBoolean("modelLoaded", landmarker != null)
        map.putString("loadedQuality", loadedQuality)
        map.putDouble("lastInferenceMs", lastInferenceMs.toDouble())
        promise.resolve(map)
    }

    /** Continuous-detection methods reserved for future live-preview
     *  use. Stub-rejected today so callers that try them get a clean
     *  "not implemented" rather than a silent no-op. Wire-up adds a
     *  CameraX preview + IMAGE_ANALYSIS use case + LiveStream
     *  RunningMode landmarker. */
    @ReactMethod
    fun startContinuousDetection(quality: String?, promise: Promise) {
        promise.reject("MP_NOT_IMPLEMENTED", "startContinuousDetection is reserved for future use")
    }

    @ReactMethod
    fun stopContinuousDetection(promise: Promise) {
        promise.resolve(null)
    }

    @ReactMethod
    fun close(promise: Promise) {
        try {
            landmarker?.close()
            landmarker = null
            promise.resolve(null)
        } catch (e: Throwable) {
            promise.reject("MP_CLOSE_FAILED", e.message ?: e.toString())
        }
    }

    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        landmarker?.close()
        landmarker = null
    }

    // ─── Result serialization ─────────────────────────────────────

    /** Translate PoseLandmarkerResult into the JS shape:
     *    {
     *      poseFound: boolean,
     *      landmarks: [{ x, y, z, visibility, presence } * 33],
     *      worldLandmarks: [...] | null,
     *      inferenceMs: number,
     *    }
     *  The visibility field is 0..1 — used by poseEstimator's
     *  joint_confidence rollup downstream. */
    private fun serializeResult(result: PoseLandmarkerResult, ms: Long): WritableMap {
        val out = Arguments.createMap()
        val landmarks: WritableArray = Arguments.createArray()
        val worldLandmarks: WritableArray = Arguments.createArray()
        var found = false
        if (result.landmarks().isNotEmpty()) {
            found = true
            for (lm in result.landmarks()[0]) {
                val p = Arguments.createMap()
                p.putDouble("x", lm.x().toDouble())
                p.putDouble("y", lm.y().toDouble())
                p.putDouble("z", lm.z().toDouble())
                p.putDouble("visibility", lm.visibility().orElse(0f).toDouble())
                p.putDouble("presence", lm.presence().orElse(0f).toDouble())
                landmarks.pushMap(p)
            }
            if (result.worldLandmarks().isNotEmpty()) {
                for (lm in result.worldLandmarks()[0]) {
                    val p = Arguments.createMap()
                    p.putDouble("x", lm.x().toDouble())
                    p.putDouble("y", lm.y().toDouble())
                    p.putDouble("z", lm.z().toDouble())
                    p.putDouble("visibility", lm.visibility().orElse(0f).toDouble())
                    p.putDouble("presence", lm.presence().orElse(0f).toDouble())
                    worldLandmarks.pushMap(p)
                }
            }
        }
        out.putBoolean("poseFound", found)
        out.putArray("landmarks", landmarks)
        out.putArray("worldLandmarks", worldLandmarks)
        out.putDouble("inferenceMs", ms.toDouble())
        return out
    }
}

File: ios-native/MediaPipePoseModule.swift
/**
 * 2026-05-23 — MediaPipe Pose Landmarker — iOS native module.
 *
 * Swift counterpart to android-native/MediaPipePoseModule.kt. Same
 * public surface; the JS bridge (services/mediaPipePoseService.ts)
 * calls into either platform identically.
 *
 * Status: SHIPS now but doesn't build on EAS until Apple Developer
 * Program enrollment lands. Once enrolled + the next EAS iOS build
 * runs, the withMediaPipePose config plugin copies this file into
 * ios/SmartPlayCaddie/MediaPipe/ and the pod 'MediaPipeTasksVision'
 * dependency resolves at pod install time.
 *
 * MediaPipe Tasks Vision iOS uses the same model file format as
 * Android (.task), so the bundled assets/mediapipe/pose_landmarker_full.task
 * is shared across both platforms — single asset, single source of
 * truth for the model.
 */

import Foundation
import UIKit
import React

// Resolved by the pod 'MediaPipeTasksVision' line added by
// withMediaPipePose.js. Xcode will mark as unresolved until the
// first iOS prebuild + pod install runs.
import MediaPipeTasksVision

@objc(MediaPipePose)
class MediaPipePose: NSObject {

    private var landmarker: PoseLandmarker?
    private var loadedQuality: String = "full"
    private var lastInferenceMs: Int = 0
    private let queue = DispatchQueue(label: "com.smartplaycaddy.mediapipe", qos: .userInitiated)

    @objc static func requiresMainQueueSetup() -> Bool { return false }

    private func modelPath(forQuality quality: String) -> String? {
        let name: String
        switch quality {
        case "lite":   name = "pose_landmarker_lite"
        case "heavy":  name = "pose_landmarker_heavy"
        default:       name = "pose_landmarker_full"
        }
        // The withMediaPipePose config plugin copies the model into
        // SmartPlayCaddie/Resources/mediapipe/<name>.task — Xcode
        // bundles that directory by default.
        return Bundle.main.path(forResource: name, ofType: "task", inDirectory: "mediapipe")
    }

    /** Lazy init. Thread-safe via the queue. */
    private func ensureLandmarker(quality: String) throws {
        if landmarker != nil && loadedQuality == quality { return }
        guard let path = modelPath(forQuality: quality) else {
            throw NSError(
                domain: "MediaPipePose",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "pose_landmarker_\(quality).task not in app bundle"],
            )
        }
        let options = PoseLandmarkerOptions()
        options.baseOptions.modelAssetPath = path
        // iOS doesn't expose a GPU/CPU delegate selector like Android;
        // the SDK chooses internally. Performance is comparable.
        options.runningMode = .image
        options.numPoses = 1
        options.minPoseDetectionConfidence = 0.4
        options.minPosePresenceConfidence = 0.4
        options.minTrackingConfidence = 0.4

        landmarker?.close()
        landmarker = try PoseLandmarker(options: options)
        loadedQuality = quality
    }

    // MARK: - Exposed methods

    @objc(detectPoseFromFrame:options:resolver:rejecter:)
    func detectPoseFromFrame(b64: NSString, options: NSDictionary?, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        queue.async {
            do {
                let quality = (options?["quality"] as? String) ?? "full"
                try self.ensureLandmarker(quality: quality)
                guard let landmarker = self.landmarker else {
                    rejecter("MP_INIT_FAILED", "Landmarker init returned nil", nil)
                    return
                }
                // Strip data: prefix if present.
                var clean = b64 as String
                if clean.hasPrefix("data:") {
                    if let comma = clean.firstIndex(of: ",") {
                        clean = String(clean[clean.index(after: comma)...])
                    }
                }
                guard let data = Data(base64Encoded: clean, options: .ignoreUnknownCharacters),
                      let image = UIImage(data: data) else {
                    rejecter("MP_DECODE_FAILED", "Could not decode base64 frame to UIImage", nil)
                    return
                }
                let mpImage: MPImage
                do {
                    mpImage = try MPImage(uiImage: image)
                } catch {
                    rejecter("MP_DECODE_FAILED", "MPImage init failed: \(error.localizedDescription)", error)
                    return
                }
                let t0 = Date().timeIntervalSince1970
                let result: PoseLandmarkerResult
                do {
                    result = try landmarker.detect(image: mpImage)
                } catch {
                    rejecter("MP_INFERENCE_FAILED", error.localizedDescription, error)
                    return
                }
                let ms = Int((Date().timeIntervalSince1970 - t0) * 1000.0)
                self.lastInferenceMs = ms
                DispatchQueue.main.async {
                    resolver(self.serializeResult(result, ms: ms))
                }
            } catch {
                rejecter("MP_INIT_FAILED", error.localizedDescription, error)
            }
        }
    }

    @objc(getStatus:rejecter:)
    func getStatus(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        resolver([
            "available": true,
            "modelLoaded": landmarker != nil,
            "loadedQuality": loadedQuality,
            "lastInferenceMs": lastInferenceMs,
        ])
    }

    @objc(startContinuousDetection:resolver:rejecter:)
    func startContinuousDetection(quality: NSString, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        rejecter("MP_NOT_IMPLEMENTED", "startContinuousDetection is reserved for future use", nil)
    }

    @objc(stopContinuousDetection:rejecter:)
    func stopContinuousDetection(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        resolver(nil)
    }

    @objc(close:rejecter:)
    func close(resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
        landmarker?.close()
        landmarker = nil
        resolver(nil)
    }

    // MARK: - Serialization

    private func serializeResult(_ result: PoseLandmarkerResult, ms: Int) -> [String: Any] {
        var landmarks: [[String: Double]] = []
        var worldLandmarks: [[String: Double]] = []
        var found = false
        if let lmList = result.landmarks.first {
            found = true
            for lm in lmList {
                landmarks.append([
                    "x": Double(lm.x),
                    "y": Double(lm.y),
                    "z": Double(lm.z),
                    "visibility": Double(lm.visibility?.floatValue ?? 0),
                    "presence":   Double(lm.presence?.floatValue ?? 0),
                ])
            }
            if let worldList = result.worldLandmarks.first {
                for lm in worldList {
                    worldLandmarks.append([
                        "x": Double(lm.x),
                        "y": Double(lm.y),
                        "z": Double(lm.z),
                        "visibility": Double(lm.visibility?.floatValue ?? 0),
                        "presence":   Double(lm.presence?.floatValue ?? 0),
                    ])
                }
            }
        }
        return [
            "poseFound": found,
            "landmarks": landmarks,
            "worldLandmarks": worldLandmarks,
            "inferenceMs": ms,
        ]
    }

    deinit {
        landmarker?.close()
    }
}
